TWEAK: Layout/styling overhaul for the F1 menu

aMannus opened the pull request:
What this PR does:

- Add helper functions for inserting padding above or below items and uses them to introduce padding around menu items.
- Adjust window padding in general
- Change sliders so they fit windows better
- Readjusted the locations of quite a few settings. For example, all D-pad related options that were under "Controller" before are now in their own flyout called "Controls" under "Enhancements".
- Consolidated "Audio", "Controller", "Graphics" and "Languages" under one menu called "Settings".
- Changed all checkboxes that open external windows to newly styled buttons. The buttons can still open and close the windows, but a button communicates the behaviour of it better than a checkbox would. Also added the ability to close the windows with the "X" button to windows that were missing it. The Cosmetics Editor got its flag removed for not taking focus on opening to elimate a double press before input is taken in the opened window.

I'd love for this to get in before the cleanup in `ImGuilmpl.cpp` is done, but I don't know if that's feasible or not.

Overview of the entire menu:
https://cdn.discordapp.com/attachments/989182407213219840/1005179298883313704/2022-08-05_20-22-07.mp4

Direct comparison to illustrate differences:

Old:
![image](https://user-images.githubusercontent.com/4244591/182711400-20e80ee7-26dc-48cf-9625-ebbbc09dbbba.png)

New:
![image](https://user-images.githubusercontent.com/4244591/183086867-cc758dfc-a9aa-4117-965f-5dba81b15e57.png)

aMannus committed:
First pass of UX changes

aMannus committed:
More padding/styling/layout

aMannus committed:
More styling

aMannus committed:
Moar styling

aMannus committed:
Some more styling

aMannus committed:
Implemented padding helpers

aMannus committed:
More styling, added closing buttons to windows

aMannus committed:
Fixed merge conflict mistake

aMannus committed:
Fixed new enhancements

aMannus committed:
Hopefully fix jenkins errors

dcvz commented:
I haven't reviewed the code yet, just done a visual review. Great work, I think it looks great!
A bit of feedback:

1. I like that options that open a window are buttons now, but the buttons kind of look like headers now, since they use a similar tint as the the currently selected dropdown. I don't know what the right solution is (probably takes some exploration) but perhaps an outline instead of solid background might make it more obvious its a clickable button?

Maybe changing the text to "Open XXXX" might already make it seem less of a header since its no longer a passive text.

<img width="224" alt="Screenshot 2022-08-04 at 23 14 06" src="https://user-images.githubusercontent.com/2475932/182953762-783cfab4-3239-4cc7-ad9f-3acaae695580.png">

2. I would change Settings > Graphics > Internal Resolution to have steppers - this is one options where being precise matters (because it affects rendering logic) and sometimes its hard to get the 100% perfectly. 

3. I much prefer buttons for new windows than the checkboxes, but it **was** nice to know which ones were open already.. now for example if I have stats on, and I open the menu I don't get any visual feedback that its open nor can I close it by clicking again, it also doesn't refocus it - nothing happens.

4. For all buttons that open windows you went with full width except the Cosmetics Editor

<img width="318" alt="Screenshot 2022-08-04 at 23 18 35" src="https://user-images.githubusercontent.com/2475932/182954556-34397e30-fc3a-4e73-a24c-603566331516.png">


Great work here! It looks really nice :)

dcvz reviewed (approved):
Looks good! I like the UI helpers you made.
Approving this though there's still pending conversations around the UI feedback

aMannus committed:
Changed button behaviour, more styling

aMannus committed:
Merge branch 'develop-zhora' into soh-f1-menu-rework

aMannus commented:
I've just pushed an update that addresses the feedback of @dcvz and also picked up more stuff that I noticed after the initial PR. Also adressed some feedback I got on discord.

The opening post is updated to reflect all the new changes.

aMannus committed:
Tiny code cleanup

aMannus committed:
Change buttons from close/open to > when open

aMannus committed:
Small button spacing fix

aMannus commented:
The buttons currently use ">" to indicate they're opened. This would be better as a checkmark. Since ImGui doesn't really support Unicode characters easily, we've talked about adding Font Awesome to use that in the future. However, this feels out of scope for this PR, and I'd rather not postpone this too much because of the incoming ImGui cleanup.

I made an issue for that here: https://github.com/HarbourMasters/Shipwright/issues/1064

I do think however, that this solution is acceptable until we get proper checkmarks (through FA) in.

PurpleHato reviewed (approved):
LGTM! Great rework!

dcvz reviewed (approved):
Great work! Not loving the current open window markers but I saw the conversation on Discord and know there's separate plans for it so I'd say LGTM 👍

Baoulettes reviewed (approved):
Great PR, just tried it and without doubt for me it look and feel better
At first the button of Cosmetics editor I was why it take full width but aligned to left, 
But once I seen that tick ( > )  when you open it that make sense.
I actually really like that approach

aMannus commented:
> Great PR, just tried it and without doubt for me it look and feel better At first the button of Cosmetics editor I was why it take full width but aligned to left, But once I seen that tick ( > ) when you open it that make sense. I actually really like that approach

Just a little clarification on the full-width part. It's also full width to bring it in line with all the other buttons that open windows. That it works well with the "tick" is honestly just a nice side benefit.

briaguya0 commented on libultraship/libultraship/ImGuiImpl.cpp:
```diff
@@ -592,31 +595,43 @@ namespace SohImGui {
             ImGui::SetCursorPosX(ImGui::GetCursorPosX() - 7.0f);
             if (ImGui::Button(PlusBTNName.c_str())) {
-                val += 0.1f;
+                if (!isPercentage)
+                    val += 0.1f;
+                else
+                    val += 0.01f;
```
didn't see this mentioned in the pr description, is this a bugfix?

aMannus replied:
Kind of. When I added the plus/minus buttons to the internal resolution slider, it would jump up by 10% at a time instead of 1%. It's just that there wasn't any float sliders before that used percentages and used the plus/minus buttons.

briaguya0 replied:
Cool!

briaguya0 commented on libultraship/libultraship/ImGuiImpl.cpp:
```diff
@@ -1522,7 +1615,7 @@ namespace SohImGui {
                     if (!isBetaQuestEnabled) {
                         ImGui::PopItemFlag();
-                        ImGui::PopStyleVar();
+                        ImGui::PopStyleVar(1);
```
what does adding the `1` here do?

aMannus replied:
If we ever apply a StyleVar in a more global fashion, StyleVars need to be popped properly one by one. Not specifying the "1" here (which is the count of the amount of StyleVars popped), it would incorrectly pop the globally applied StyleVars as well. In this case, it's popping the StyleVar from this:

```
if (!isBetaQuestEnabled) {
                        ImGui::PushItemFlag(ImGuiItemFlags_Disabled, true);
                        ImGui::PushStyleVar(ImGuiStyleVar_Alpha, ImGui::GetStyle().Alpha * 0.5f);
                    }
```

briaguya0 replied:
Sounds good

briaguya0 commented on libultraship/libultraship/ImGuiImpl.cpp:
```diff
@@ -1592,10 +1744,14 @@ namespace SohImGui {
             const float framerate = ImGui::GetIO().Framerate;
             ImGui::PushStyleColor(ImGuiCol_Border, ImVec4(0, 0, 0, 0));
-            ImGui::Begin("Debug Stats", nullptr, ImGuiWindowFlags_NoFocusOnAppearing);
+            ImGui::Begin("Debug Stats", &statsWindowOpen, ImGuiWindowFlags_NoFocusOnAppearing);
```
might be nice to add something to the PR description explaining this change

aMannus replied:
It's this part from the OP: `Also added the ability to close the windows with the "X" button to windows that were missing it.`

briaguya0 replied:
coolcool, makes sense!

briaguya0 reviewed (approved):
overall it's looking good, noticed a few non-styling changes that aren't noted in the PR description and would be good to have documented, but nothing that makes me think we shouldn't :shipit:

briaguya0 commented on libultraship/libultraship/ImGuiImpl.cpp:
```diff
@@ -1465,7 +1558,7 @@ namespace SohImGui {
+                    PaddedEnhancementCheckbox("Enable Beta Quest", "gEnableBetaQuest", true, false);
                     Tooltip("Turns on OoT Beta Quest. *WARNING* This will reset your game.");
                     betaQuestEnabled = CVar_GetS32("gEnableBetaQuest", 0);
                     if (betaQuestEnabled) {
```
could you remove
```
ImGui::Text("After Slider Beta Quest World: %d", betaQuestWorld);
```
and 
```
ImGui::Text("After Clamp Beta Quest World: %d", betaQuestWorld);
```

they were left in on accident and i figure just removing them with this PR will be easier than dealing with a merge conflict

aMannus replied:
There's more merge conflicts than just that, so I'll just try and get the latest zhora merged into this instead.

briaguya0 replied:
yes, but that hasn't been updated in latest zhora, the idea was to hold off on adding more to your imgui merge plate by waiting for your PR to fix that (but if you can just get it added that'd be perfect)

garrettjoecox replied:
You probably don't need to handle this here, but could be low hanging fruit #1103

aMannus replied:
> You probably don't need to handle this here, but could be low hanging fruit #1103

I'd prefer to pick this up in a seperate PR if neccesary. I just want this to be in sooner rather than later because merge conflicts are getting really hard to sift through.

aMannus replied:
@briaguya-ai Text is removed in latest commit.

aMannus committed:
Merge branch 'develop-zhora' into soh-f1-menu-rework

aMannus committed:
Small styling changes after merge

aMannus committed:
Small fix after merge mistake

aMannus commented:
Waiting for Jenkins to say it's happy, but it should be good to go again.

That said, I'll have to follow up with another PR to include options in the enhancement presets that were moved to "Enhancements" in this PR. There was some new options added since then I'd have to include anyway.

I'll see if I can pick up https://github.com/HarbourMasters/Shipwright/issues/1103 in that PR as well.
